app/frontend/src/features/profile/actions/FriendActions.tsx

```diff
@@ -0,0 +1,47 @@
+import { CircularProgress } from "@material-ui/core";
+import { SetMutationError } from "features/connections/friends";
+import AddFriendButton from "features/connections/friends/AddFriendButton";
+import useFriendList from "features/connections/friends/useFriendList";
+import useFriendRequests from "features/connections/friends/useFriendRequests";
+import MessageUserButton from "features/profile/actions/MessageUserButton";
+import PendingFriendReqButton from "features/profile/actions/PendingFriendReqButton";
+import { User } from "pb/api_pb";
+import React from "react";
+
+interface FriendActionsProps {
+  user: User.AsObject;
+  setMutationError: SetMutationError;
+}
+
+export default function FriendActions({
+  user,
+  setMutationError,
+}: FriendActionsProps) {
+  const { isLoading: isFriendRequestsLoading } = useFriendRequests("received");
+  const { isLoading: isFriendsLoading } = useFriendList();
```

Comment:
It's better for the component that uses the data to display loading indicators. They can have better styling more similar to the end result.

Reply:
I see what you mean, but it won't work for two reasons. Firstly you can't rely on queries to actually query new data - they might return something from the cache without fetching if it's still considered fresh.  (We have set users to stay fresh for 5 minutes I think). If it needs to be re-fetched, there needs to be invalidation somewhere. 

Secondly, it's too opaque - this component decides what to render from `user.friends`, making it responsible for having an up to date user object is mixing responsibilities, which could lead to difficult-to-trace bugs.

In any case, I don't think it's actually needed now since you added invalidating the user on the successful response.

Reply:
Makes sense! I've removed it.